storage: use new `pebble.CloneOptions` when cloning iterators

erikgrinaker opened the pull request:
This patch avoids first cloning and then reconfiguring the cloned
iterator, by constructing the clone with the correct settings in the
first place.

This shows a slight improvement on `MVCCGet` benchmarks that were
modified to always use cloned iterators. This would also happen
normally in many cases (e.g. when using an intent interleaving iterator
with a raw `Engine`, or when constructing a time-bound iterator which
requires two iterators from the underlying reader).

```
MVCCGet_Pebble/batch=false/versions=1/valueSize=8-24      4.87µs ± 3%    4.82µs ± 2%    ~     (p=0.101 n=10+10)
MVCCGet_Pebble/batch=false/versions=10/valueSize=8-24     5.88µs ± 1%    5.83µs ± 0%  -0.94%  (p=0.004 n=10+9)
MVCCGet_Pebble/batch=false/versions=100/valueSize=8-24    13.7µs ± 6%    13.7µs ± 4%    ~     (p=0.971 n=10+10)
MVCCGet_Pebble/batch=true/versions=1/valueSize=8-24       2.76µs ± 1%    2.76µs ± 1%    ~     (p=0.871 n=10+10)
MVCCGet_Pebble/batch=true/versions=10/valueSize=8-24      3.89µs ± 1%    3.86µs ± 2%  -0.76%  (p=0.011 n=10+10)
MVCCGet_Pebble/batch=true/versions=100/valueSize=8-24     10.2µs ± 3%    10.2µs ± 4%    ~     (p=0.912 n=10+10)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82981)
<!-- Reviewable:end -->

erikgrinaker committed:
storage: use new `pebble.CloneOptions`

This patch avoids first cloning and then reconfiguring the cloned
iterator, by constructing the clone with the correct settings in the
first place.

This shows a slight improvement on `MVCCGet` benchmarks that were
modified to always use cloned iterators. This would also happen
normally in many cases (e.g. when using an intent interleaving iterator
with a raw `Engine`, or when constructing a time-bound iterator which
requires two iterators from the underlying reader).

```
MVCCGet_Pebble/batch=false/versions=1/valueSize=8-24      4.87µs ± 3%    4.82µs ± 2%    ~     (p=0.101 n=10+10)
MVCCGet_Pebble/batch=false/versions=10/valueSize=8-24     5.88µs ± 1%    5.83µs ± 0%  -0.94%  (p=0.004 n=10+9)
MVCCGet_Pebble/batch=false/versions=100/valueSize=8-24    13.7µs ± 6%    13.7µs ± 4%    ~     (p=0.971 n=10+10)
MVCCGet_Pebble/batch=true/versions=1/valueSize=8-24       2.76µs ± 1%    2.76µs ± 1%    ~     (p=0.871 n=10+10)
MVCCGet_Pebble/batch=true/versions=10/valueSize=8-24      3.89µs ± 1%    3.86µs ± 2%  -0.76%  (p=0.011 n=10+10)
MVCCGet_Pebble/batch=true/versions=100/valueSize=8-24     10.2µs ± 3%    10.2µs ± 4%    ~     (p=0.912 n=10+10)
```

Release note: None

jbowens reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

the new `CloneOptions` (and the fix for the bug that you ran into when testing this previously) have landed on `master`.

> Reviewed 2 of 5 files at r1, 2 of 3 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82981#-:-N5pOD1-59pMj8hPTaZr:b-9d0rhn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

nicktrav reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :shipit: 

> Reviewed 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82981#-:-N6jfJ-KBnuxZ1m8WxML:byzm7aw)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->
